iceoryx_platform/generic/include/iceoryx_platform/atomic.hpp

```diff
@@ -0,0 +1,518 @@
+#include <type_traits>
+
+namespace iox::concurrent
+{
+/// @brief An alias to the std::atomic_flag
+using AtomicFlag = std::atomic_flag;
+
+/// @brief A thin wrapper for a 'std::atomic' which ensures that all atomic operations are always lock-free in order to
+/// ensure safe usage across processes in shared memory. For detailed documentation please have a look at
+/// https://en.cppreference.com/w/cpp/atomic/atomic
+template <typename T>
+class Atomic
+{
+  private:
+    template <typename U>
+    using enable_if_integral_t = std::enable_if_t<std::is_same_v<T, U> && std::is_integral_v<U>, T>;
+
+    template <typename U>
+    using enable_if_pointer_t = std::enable_if_t<std::is_same_v<T, U> && std::is_pointer_v<U>, T>;
+
+    template <typename U>
+    using enable_if_integral_or_pointer_t =
+        std::enable_if_t<std::is_same_v<T, U> && (std::is_integral_v<U> || std::is_pointer_v<U>), T>;
+
+  public:
+    static_assert(std::atomic<T>::is_always_lock_free,
+                  "The 'iox::Atomic' must work across process boundaries and must therefore be always lock-free!");
+
+    /// @brief This is always 'true' for 'iox::Atomic'
+    static constexpr bool is_always_lock_free = std::atomic<T>::is_always_lock_free;
+
+    /// @brief Constructs a new 'iox::Atomic' with a default value
+    constexpr Atomic() noexcept
+        : m_value{T()}
+    {
+    }
+
+    /// @brief Constructs a new 'iox::Atomic' with the given value
+    explicit constexpr Atomic(T value) noexcept
+        : m_value{value}
+    {
+    }
+
+    /// @brief Similar to the std::atomic, the 'iox::Atomic' is not copy constructible
+    Atomic(const Atomic& other) = delete;
+    /// @brief Similar to the std::atomic, the 'iox::Atomic' is not copy assignable
+    Atomic& operator=(const Atomic&) = delete;
+
+    Atomic(Atomic&& rhs) noexcept = default;
+    Atomic& operator=(Atomic&& rhs) noexcept = default;
+
+    ~Atomic() = default;
+
+    /// @brief Atomically assigns the given value to the 'iox::Atomic' and returns the given value. Equivalent to
+    /// 'store(value)'
+    // NOLINTNEXTLINE(cppcoreguidelines-c-copy-assignment-signature) This mimics the std::atomic copy assignment operator
+    T operator=(T value) noexcept
+    {
+        m_value = value;
+        return value;
+    }
+
+    /// @brief Atomically assigns the given value to the 'iox::Atomic' and returns the given value. Equivalent to
+    /// 'store(value)'
+    // NOLINTNEXTLINE(cppcoreguidelines-c-copy-assignment-signature) This mimics the std::atomic copy assignment operator
+    T operator=(T value) volatile noexcept
+    {
+        m_value = value;
+        return value;
+    }
+
+    /// @brief Atomically loads and returns the stored value. Equivalent to 'load()'
+    explicit operator T() const noexcept
+    {
+        return m_value.operator T();
+    }
+
+    /// @brief Atomically loads and returns the stored value. Equivalent to 'load()'
+    explicit operator T() const volatile noexcept
```

Comment:
This is a deviation from std::atomic but clang-tidy complains when it is not explicit